Add support for containers of move only typed values to range.

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier commented on unit/util/range.cpp:
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
+      REQUIRE(both_inputs[19].value == 10);
+    }
+  }
+}
```
:pick: missing new line

thk123 replied:
:no_entry_sign: missing new line

romainbrenguier commented on unit/util/range.cpp:
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
+      input1.emplace_back(i);
+    std::vector<move_onlyt> input2;
+    for(int i = 1; i <= 10; ++i)
+      input2.emplace_back(i);
```
:pick: input2 is not used in all the tests, consider moving it closer to the test actually using it (maybe in a second GIVEN)

thk123 commented on unit/util/range.cpp:
```diff
@@ -65,3 +65,95 @@ SCENARIO("range tests", "[core][util][range]")
+  move_onlyt(const move_onlyt &) = delete;
+  move_onlyt &operator=(const move_onlyt &) = delete;
+
+  explicit move_onlyt(int value) : value{value} {};
```
:pick: I'd have probably done an actually move-only type (e.g. with a `std::unique_ptr`) so I'd be sure that I'd really fixed it.

thomasspriggs replied:
'move_onlyt' is an actual move only type, because I have deleted to copy constructor and the copy assignment operator. Attempting to copy it will result in a compile error.

thk123 commented on src/util/range.h:
```diff
@@ -142,6 +142,16 @@ class filter_iteratort
     return tmp;
   }
 
+  value_type &operator*()
```
:question: The commit message + change seem unrelated to me, perhaps you could expand in the commit message why these are required.

thk123 commented on src/util/range.h:
```diff
@@ -244,6 +244,20 @@ struct concat_iteratort
     return tmp;
   }
 
+  value_type &operator*()
```
:question: Ditto - please elaborate.

thk123 reviewed (commented):
Could the commit messages contain a little more info why these things are required.

tautschnig commented on src/util/range.h:
```diff
@@ -357,12 +357,11 @@ ranget<iteratort> make_range(iteratort begin, iteratort end)
   return ranget<iteratort>(begin, end);
 }
 
```
Comment on the commit message: s/in oder/in order/

tautschnig reviewed (approved):
Some minor bits left as noted by reviewers.

thomasspriggs committed:
Allow `make_range` to construct a range of non-const values

This is required in order to allow values in the constructed range to be
moved from.

thomasspriggs committed:
Add non-const overloads of the `*` and `->` for `filter_iteratort`

This adds support for moving from the values accessible through the
result of `ranget.filter`.

thomasspriggs committed:
Add non-const overloads of the `*` and `->` for `concat_iteratort`

This adds support for moving from the values accessible through the
result of `ranget.concat`.

thomasspriggs committed:
Update `map_iteratort` to use a `shared_ptr` for `current`

This allows mapping into move only types. This was not previously
possible, because the copy constructor of `map_iteratort` copied the
value pointed to by its unique pointer in its `current` field. This can
also be considered to be more correct, because copying an iterator
should be expected to yield two copies, which point to the same value,
rather than two copies, which point to two separate values.

thomasspriggs committed:
Add non-const overloads of the `*` and `->` for `map_iteratort`

This adds support for moving from the values accessible through the
result of `ranget.map`.

thomasspriggs committed:
Add unit test on ranges of move only values

thomasspriggs commented:
I have addressed the review comments. I have also added better move support and corresponding unit tests for the `map` operation, as I realised I had previously missed this.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cdcd544d582aa877ffc43948315be5c4b647f990).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95016060
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thomasspriggs commented:
I have rebased in order to re-check test-gen compatibility.

thk123 commented on unit/util/range.cpp:
```diff
@@ -65,3 +65,121 @@ SCENARIO("range tests", "[core][util][range]")
+  {
+    const std::vector<int> input{1, 2, 3, 4, 5};
+
+    THEN("The vector can be mapped into a range of move-only types")
```
If it works - you should add a test that maps from a move only type to a regular type, and one that moves the same type through the map

thomasspriggs replied:
The test which "maps from a move only type to a regular type" already exists. As we discussed, you missed it when reviewing. The `f` passed into `map` can't move from its parameter, because the parameter has to be a const reference. I have expanded the doxygen to document the extent of the support for move-only types and its limitations.

thk123 reviewed (commented):
New changes look good and previous comments addressed - if possible I'd like to see two more tests for `map`

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0b73746794271d3276a8ba0df02484f3d5657756).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95095292

thk123 commented:
Also @romainbrenguier should definitely re-review the unique_ptr -> shared_ptr before merging.

thomasspriggs committed:
Expand doxygen on `ranget.map`

The new doxygen explains the extent and limitations of `ranget.map`'s
support of move-only types.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bbc201668e32867361841f2ad4eba3ecafe3185b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95102012

romainbrenguier reviewed (approved):
The change from unique to shared pointer looks good
